src/utils.ts

```diff
@@ -271,3 +271,28 @@ export const convertOrderByToMetricId = (orderBy: string): string | undefined =>
   const metricIdMatches = orderBy.match(/^(\d+)/);
   return metricIdMatches?.[1];
 };
+
+export function memoizeAsync<Args extends any[], Result>(
+  fn: (...args: Args) => Promise<Result>,
+  getKey: (...args: Args) => string = (...args) => JSON.stringify(args)
+): (...args: Args) => Promise<Result> {
+  const cache = new Map<string, Promise<Result>>();
+
+  return (...args: Args): Promise<Result> => {
+    const key = getKey(...args);
+
+    if (cache.has(key)) {
+      return cache.get(key)!;
+    }
+
+    const resultPromise = fn(...args);
+    cache.set(key, resultPromise);
```

Comment:
caching the promise may leak memory. how about we cache the returned value instead of the promise?

Reply:
Good call, changed so it caches the result